receiver/sapmreceiver/trace_receiver.go

```diff
@@ -0,0 +1,239 @@
+	"github.com/gorilla/mux"
+	"github.com/open-telemetry/opentelemetry-collector/consumer"
+	"github.com/open-telemetry/opentelemetry-collector/observability"
+	"github.com/open-telemetry/opentelemetry-collector/oterr"
+	"github.com/open-telemetry/opentelemetry-collector/receiver"
+	jaegertranslator "github.com/open-telemetry/opentelemetry-collector/translator/trace/jaeger"
+	splunksapm "github.com/signalfx/sapm-proto/gen"
+	"github.com/signalfx/sapm-proto/sapmrequestparser"
+	"go.opencensus.io/trace"
+	"go.uber.org/zap"
+)
+
+const (
+	// TraceEndpointV2 is the endpoint to receive sapm on
+	TraceEndpointV2 = "/v2/trace"
+
+	traceSource string = "sapm"
+
+	contentTypeHeader = "Content-Type"
+	xprotobuf         = "application/x-protobuf"
+
+	acceptEncodingHeader  = "Accept-Encoding"
+	contentEncodingHeader = "Content-Encoding"
+	gzipEncoding          = "gzip"
+)
+
+var gzipWriterPool = &sync.Pool{
+	New: func() interface{} {
+		return gzip.NewWriter(ioutil.Discard)
+	},
+}
+
+// sapmReceiver receives spans in the Splunk SAPM format over HTTP
+type sapmReceiver struct {
+	// mu protects the fields of this type
+	mu        sync.Mutex
+	startOnce sync.Once
+	stopOnce  sync.Once
+	logger    *zap.Logger
+
+	config *Config
+	server *http.Server
+
+	nextConsumer consumer.TraceConsumer
+}
+
+// handleRequest parses an http request containing sapm and passes the trace data to the next consumer
+func (sr *sapmReceiver) handleRequest(ctx context.Context, req *http.Request) error {
+	sapm, err := sapmrequestparser.ParseTraceV2Request(req)
+	// errors processing the request should return http.StatusBadRequest
+	if err != nil {
+		return err
+	}
+
+	// pass the SAPM and error to the receiver function
+	// process the sapm
+	if sapm == nil {
+		return errors.New("incoming sapm was nil")
+	}
+
+	// process sapm batches
+	for _, batch := range sapm.Batches {
+		// convert the jager batches to OCProto
+		td, err := jaegertranslator.ProtoBatchToOCProto(*batch)
+		if err != nil {
+			return err
+		}
+		td.SourceFormat = "sapm"
+
+		// pass the trace data to the next consumer
+		err = sr.nextConsumer.ConsumeTraceData(ctx, td)
+		if err != nil {
+			return fmt.Errorf("error passing trace data to next consumer: %v", err.Error())
+		}
+
+		// We MUST unconditionally record metrics from this reception.
+		observability.RecordMetricsForTraceReceiver(ctx, len(batch.Spans), len(batch.Spans)-len(td.Spans))
+	}
+
+	return nil
+}
+
+// HTTPHandlerFunction returns an http.HandlerFunc that handles SAPM requests
+func (sr *sapmReceiver) HTTPHandlerFunc(rw http.ResponseWriter, req *http.Request) {
+	// create context with the receiver name from the request context
+	ctx := observability.ContextWithReceiverName(req.Context(), "sapm")
+
+	// trace this request
+	_, span := trace.StartSpan(ctx, traceSource)
+	defer span.End()
+
+	// handle the request payload
+	err := sr.handleRequest(ctx, req)
+	if err != nil {
+		// TODO account for this error (throttled logging or metrics)
```

Comment:
Will do.  @tigrannajaryan is there some kind of convention in otel for handling this?  I don't think zap has a rate limited logger (at least i couldn't find one in their docs).  I hesitate to put a plain log statement there because log files could blow up from a bad actor.

Reply:
hrm I think i found something in zap. They refer to this as "sampling".  I'll have to read their doc further to understand how to use that feature.  I've never used zap before.

Reply:
I think it is best to record a metric instead of logging. I believe that is how we expose most of bad things happening in Collector that can happen because of input data.